src/http/server.rs

```diff
@@ -81,3 +96,42 @@ pub async fn start_server(config: Config) -> Result<()> {
 
   Ok(server.await.map_err(CLIError::from)?)
 }
+
+#[async_trait::async_trait]
+pub trait RequestExecutor {
+  #[allow(clippy::too_many_arguments)]
+  async fn execute(
+    &self,
+    bytes: &hyper::body::Bytes,
+    req_ctx: Arc<RequestContext>,
+    server_ctx: &ServerContext,
+  ) -> Result<GraphQLResponse>;
+}
+
+pub struct SingleRequestExecutor {}
+#[async_trait::async_trait]
+impl RequestExecutor for SingleRequestExecutor {
+  async fn execute(
+    &self,
+    bytes: &hyper::body::Bytes,
+    req_ctx: Arc<RequestContext>,
+    server_ctx: &ServerContext,
+  ) -> Result<GraphQLResponse> {
+    let request: async_graphql_hyper::GraphQLRequest = serde_json::from_slice(bytes)?;
+    Ok(request.data(req_ctx.clone()).execute(&server_ctx.schema).await)
+  }
+}
+
+pub struct BatchRequestExecutor {}
+#[async_trait::async_trait]
+impl RequestExecutor for BatchRequestExecutor {
+  async fn execute(
+    &self,
+    bytes: &hyper::body::Bytes,
+    req_ctx: Arc<RequestContext>,
+    server_ctx: &ServerContext,
+  ) -> Result<GraphQLResponse> {
+    let request: async_graphql_hyper::GraphQLBatchRequest = serde_json::from_slice(bytes)?;
+    Ok(request.data(req_ctx.clone()).execute(&server_ctx.schema).await)
+  }
+}
```

Comment:
nitpick
Consider moving request_handle to it's own file. The logic has become quite complicated now.